test/nodes/test_shaper.py

```diff
@@ -0,0 +1,563 @@
+from haystack import Pipeline, Document
+
+
+def test_basic_function_invocation(tmp_path):
+    with open(tmp_path / "tmp_config.yml", "w") as tmp_file:
+        tmp_file.write(
+            f"""
+            version: ignore
+            components:
+            - name: shaper
+              params:
+                inputs:
+                    query:
+                      func: expand
+                      output: questions
+                      params:
+                        expand_target: query
+                        size:
+                          func: len
+                          params:
+                            - documents
+                    documents:
+                      func: concat_docs
+                      output: documents
```

Comment:
I do find it a bit odd that if we specify `output: documents` to be the same as the input name (in this case also `documents`) then the result will return like 
```python
{'query': ..., 'documents': 'Berlin is an amazing city. I love Berlin', ... }
```
with the `documents` having been overwritten. 

However, if I specify a new output name (like `output: docs`) then the variable is saved in the `meta` field of result, which looks like 
```python
{'query': ..., 'documents': [<Document: {...}>, <Document: {...}>],
'meta': {'invocation_context': {'questions': ['What can you tell me about Berlin?', 'What can you tell me about Berlin?'], 'docs': 'Berlin is an amazing city. I love Berlin'}}
```
Is there a reason we need to put it into the `invocation_context` instead of the same level as the keys `query` and `documents`?

Reply:
Yes, there is a reason, and that is my biggest pain point ever :-) The constrained signature of the run method, it doesn't have kwargs ...ugh.

Reply:
Ahh okay, I see.